### ISSUE TITLE
[meta] Split the core

### ISSUE THREAD

**Issue by Hywan** [hoaproject/Core#96]
### History

The core gathers several thin layers into a single library for performance reasons. Let me explain it shortly. Parsing time is not very costly compared to I/O: Accessing to a file, opening it and reading it. The core aimed at putting the minimum always needed codes into small files, in order to reduce this I/O time. Since recent PHP versions and recent I/O progresses (SSD, kernel, FS…), this is no longer necessary. We can then split the core.

### Global idea

The core contains the following layers:
* `Hoa\Core\Consistency`,
* `Hoa\Core\Data`,
* `Hoa\Core\Event`,
* `Hoa\Core\Exception`,
* `Hoa\Core\Parameter`,
* `Hoa\Core\Protocol`.

We can split it into the following libraries:
* `Hoa\Consistency`,
* `Hoa\Event`,
* `Hoa\Exception`,
* `Hoa\Protocol`.

They can live outside the core without any problem. So far, though, `Hoa\Core\Data` and `Hoa\Core\Parameter` will be kept as is.

### Dealing with pre-loading

`Hoa\Core\Consistency`, `Hoa\Core\Parameter` and `Hoa\Core\Protocol` were *de-facto* pre-loaded with `Hoa\Core\Core`. This latter is automatically pre-loaded with the `autoload.files` Composer's property. Whatever the split, we have to ensure this behavior.

#### Case of `Hoa\Core\Consistency`

This layer contains:
* An autoloader,
* An xcallable implementation and,
* Consistency helpers/tools accross PHP VMs (backport features to prior VMs, naming consistencies, feature consistencues etc.).

Only the latter is important at any runtime and will be loaded as a separated file with Composer. This will reduce the number of loaded code by approximately 60%.

#### Case of `Hoa\Core\Parameter`

Will not be extracted yet. Nothing to do.

#### Case of `Hoa\Core\Protocol`

We need to register the `hoa://` protocol at any runtime. We use it everywhere. It has never hurt anyone in the whole Hoa's history. So we can load the file that is responsible to register the stream wrapper with Composer. This will reduce the number of loaded code by approximately 50%.

### Steps

Here are the following steps to follow (the order is important):
- [x] `Hoa\Exception`:
  - [x] Create `Hoa\Exception` (from `Hoa\Core\Exception`),
  - [x] Write test suites,
  - [x] Update all libraries to use `Hoa\Exception`,
  - [x] Remove `Hoa\Core\Exception`.
- [x] `Hoa\Event`:
  - [x] Create `Hoa\Event` (from `Hoa\Core\Event`),
  - [x] Write test suites,
  - [x] Update all libraries  to use `Hoa\Event`,
  - [x] Remove `Hoa\Core\Event`.
- [x] `Hoa\Protocol`:
  - [x] Create `Hoa\Protocol`,
  - [x] Write test suites,
  - [x] Update all libraries to use `Hoa\Protocol`,
  - [x] Remove `Hoa\Core\Protocol`.
- [x] `Hoa\Consistency`:
  - [x] Create `Hoa\Consistency`,
  - [x] Write test suites,
  - [x] Update all libraries to use `Hoa\Consistency`,
  - [x] Remove `Hoa\Core\Consistency`.

### Next steps

~~Continue to split the core, see https://github.com/hoaproject/Core/issues/97.~~

Here are the following steps to follow:
- [x] `Hoa\Xyl\Data`:
  - [x] Move `Hoa\Core\Data` to `Hoa\Xyl\Data`,
  - [x] Update `Hoa\View` and other libraries to understand a mixed data structured instead (to remove a dependency),
  - [x] Remove `Hoa\Core\Data`.
- [x] `Hoa\Parameter`:
  - [x] Create `Hoa\Parameter` (from `Hoa\Core\Parameter`),
  - [x] Write test suites (optional so far),
  - [x] Remove `Hoa\Core\Parameter`.
- [x] Remove `Hoa\Core`:
  - [x] Mark the repository as deprecated,
  - [x] Do we empty the repository?

### Update all `composer.json` files

Because we are dropping PHP5.4 too.

  * [x] Acl
  * [x] Bench
  * [x] Cache
  * [x] Cli
  * [x] Compiler
  * [x] Consistency
  * [x] Console
  * [x] Database
  * [x] Devtools
  * [x] Dispatcher
  * [x] Dns
  * [x] Event
  * [x] Eventsource
  * [x] Exception
  * [x] Fastcgi
  * [x] File
  * [x] Graph
  * [x] Http
  * [x] Irc
  * [x] Iterator
  * [x] Json
  * [x] Locale
  * [x] Log
  * [x] Mail
  * [x] Math
  * [x] Memory
  * [x] Mime
  * [x] Model
  * [x] Notification
  * [x] Praspel
  * [x] Promise
  * [x] Protocol
  * [x] Prototype
  * [x] Realdom
  * [x] Regex
  * [x] Registry
  * [x] Router
  * [x] Ruler
  * [x] Serialize
  * [x] Session
  * [x] Socket
  * [x] Stream
  * [x] String
  * [x] Stringbuffer
  * [x] Test
  * [x] Translate
  * [x] Tree
  * [x] Ustring
  * [x] View
  * [x] Visitor
  * [x] Websocket
  * [x] Worker
  * [x] Xml
  * [x] Xmlrpc
  * [x] Xyl
  * [x] Zformat
  * [x] Zombie
  * [x]

**Comment by Hywan** [comment]
Thoughts @hoaproject/hoackers?

**Comment by Jir4** [comment]
And what happens to `Hoa\Core\Data` ? If i understand, it is kept in `Hoa\Core`, why ?

**Comment by Hywan** [comment]
@Jir4 Because I am not sure about what to do with this one so far.

**Comment by osaris** [comment]
That make sense, it would be very interesting to do some benchmarks before/after core splitting to check the real difference on performances (not saying that it'll decrease performances but numbers/data are interesting).

**Comment by Jir4** [comment]
@osaris  :+1:

**Comment by Hywan** [comment]
@osaris Also, something else must be quantified but this is really hard: The psychological impact of having a “core” for a library. This is something really important. Several people stops at this. The goal is to really continue to split. Please, see https://github.com/hoaproject/Core/issues/97.

**Comment by Hywan** [comment]
I started to split `Hoa\Core\Exception` and `Hoa\Core\Event` yesterday night.

**Comment by Hywan** [comment]
`Hoa\Event` is almost over.
`Hoa\Exception` is under splitting (the repositories have been created).

**Comment by Hywan** [comment]
I am starting to update libraries with `Hoa\Event` and `Hoa\Exception`.

**Comment by Hywan** [comment]
First run. I will continue today if I have time.

**Comment by Metalaka** [comment]
@Hywan Sorry but most of your PR introduce a parse error: https://3v4l.org/vB5Dt

```php
namespace Hoa\Exception {
    class Exception {}
}

namespace Hoa\Database {
    
    use Hoa\Exception;
    
    class Exception {}
    // Fatal error: Cannot declare class Hoa\Database\Exception because the name is already in use…
}
```

**Comment by Metalaka** [comment]
Your should probably remove the `use` statement.

```php
namespace Hoa\Database;

class Exception extends \Hoa\Exception {}

```

**Comment by Hywan** [comment]
Damned, I tested it on PHP7-nightly and it worked :-/.

**Comment by Hywan** [comment]
@Metalaka Fixed.

**Comment by Metalaka** [comment]
Is it always possible to use flex entity ?
```php
namespace Hoa\Database;

use Hoa\Exception as HoaException;

class Exception extends HoaException
{
}
```

**Comment by Hywan** [comment]
@Metalaka I think so.

**Comment by Metalaka** [comment]
@Hywan flex entity seem to work at least for `Hoa\Database`.

**Comment by Hywan** [comment]
@Metalaka Hmm, yes, I shoud have update with the shortest name. Do you think this is required?

**Comment by Hywan** [comment]
(anyway, I started to update the code).

**Comment by Metalaka** [comment]
The most important is do the same things everywhere (alias or not alias).

**Comment by Hywan** [comment]
@Metalaka We use the shortest everywhere. I started to move every library to `Hoa\Event`. I am updating this issue too.

**Comment by Hywan** [comment]
All libraries have been updated to `Hoa\Event`.

**Comment by Hywan** [comment]
All libraries have been updated to `Hoa\Exception` (final).

**Comment by Hywan** [comment]
[`Hoa\Consistency`](https://github.com/hoaproject/Consistency) has been created. I will move the code tomorrow. Hope to start test suites tomorrow too.

**Comment by Hywan** [comment]
`Hoa\Consistency` is OK. Now I will start updating other libraries.

**Comment by Hywan** [comment]
All tests are green for `Hoa\Protocol` split. Now I have to write the test suite of `Hoa\Protocol` but the integration of the new library is fine!

**Comment by Hywan** [comment]
Everything is OK!
I am starting #97.

**Comment by Hywan** [comment]
@hoaproject/hoackers Please, I need your help!
1. What about `Hoa\Zformat` instead of `Hoa\Parameter`? Do you think this is better?
2. What to do with `Hoa\Core`? See the “Next steps” Section in https://github.com/hoaproject/Core/issues/96#issue-111106258 for the questions.

Thanks!

**Comment by Jir4** [comment]
1. Why ZFormat ? Meaning ?

2. I think we should delete `Hoa\Core`, the `Core` name scares many users imo, they just wanna use a specific lib but the `Core` dependancy seems imply that the lib need many others to work

**Comment by Hywan** [comment]
@Jir4 
1. zFormat because this is the format we use to express parameters.
2. Yes, we will “remove” it, but for legacy code, it must still be present as a repository. We can mark it as deprecated and keep the code as is. Thoughts?

**Comment by Jir4** [comment]
1. Ok so :+1: 
2. So we keep it just to avoid a BC break and we will remove it when another BC break occurred ?

**Comment by Hywan** [comment]
@Jir4 We will remove it when we will no longer have any downloads. Until we have downloads, we must keep the repository live :-].

**Comment by CircleCode** [comment]
@Hywan the downside of this approach is that we will have downloads (at least for CI systems) so long as it will be available…
IMHO we should announce an estimated end-of-life

**Comment by Hywan** [comment]
@CircleCode Hmm… Every year, we can clean deprecated repositories? So far, we only have `Hoa\String` (see git.hoa-project.net). How long is enought?

**Comment by Jir4** [comment]
Several month but not years, we just have to let the time to users to update their legacy code.

**Comment by Hywan** [comment]
```
Success (95 tests, 800/800 methods, 0 void method, 0 skipped method, 180913 assertions)!
```
`Hoa\Zformat` has been created. I am removing `hoa/core` from everywhere :-).

**Comment by Hywan** [comment]
All libraries are updated. Now I am running `devtools:cs` to update the year.

**Comment by Hywan** [comment]
Done.

**Comment by Hywan** [comment]
Next step is to merge everything and create new snapshots for all libraries. I don't have enough time to do it well if something failed right now. Let see tonight or tomorrow!

**Comment by Hywan** [comment]
Done \o/.
Congratulations to everyone :fireworks: :tada: :balloon:!!

**Comment by Jir4** [comment]
:+1:

**Comment by Hywan** [comment]
:sweat_smile:

**Comment by 1e1** [comment]
👏
Great job! 
I've just read the discussion, do you know what you earn by registering `hoa://` in Composer? For projects like the official website, Rulerz, some Liip projects…

**Comment by Hywan** [comment]
@1e1 What we earn? I don't understand.

**Comment by Jir4** [comment]
I suppose that he speak about performance

**Comment by Hywan** [comment]
That's very light. Let me try a small benchmark.

**Comment by 1e1** [comment]
Yeap! (Sorry)

**Comment by Hywan** [comment]
### tl;dr

Here is small benchmark result when loading Hoa (because we have 1 require autoloaded file):

| vs | Time (for 1000 runs) | Memory |
| --- | --- | --- |
| `Hoa\Core` | 18.084s | 335408b |
| `Hoa\Consistency` | 15.959s | 56744b |
| Gain | **11.751%** |  **83.082%** |

### Time

First script `/tmp/a.sh`:

``` sh
for i in `seq 1 1000`; do
    php -r 'require "/usr/local/lib/Hoa/Core/Core.php";'
done
```

Second script `/tmp/b.sh`:

``` sh
for i in `seq 1 1000`; do
    php -r 'require "/usr/local/lib/Hoa/Consistency/Prelude.php";'
done
```

Time for `time`:

``` sh
$ time /tmp/a.sh
/tmp/a.sh  11.37s user 5.64s system 94% cpu 18.084 total

$ time /tmp/b.sh
/tmp/b.sh  9.56s user 5.07s system 91% cpu 15.959 total
```

That's an improvement of **11.751%** at loading time.

### Memory

First script `/tmp/a.php`:

``` php
<?php

$a = memory_get_usage();

require '/usr/local/lib/Hoa/Core/Core.php';

var_dump(memory_get_usage() - $a);
```

Second script `/tmp/b.php`:

``` php
<?php

$a = memory_get_usage();

require '/usr/local/lib/Hoa/Consistency/Prelude.php';

var_dump(memory_get_usage() - $a);
```

Time to run:

``` sh
$ php /tmp/a.php
int(335408)

$ php /tmp/b.php
int(56744)
```

That's an improvement of 278664b, be **272Kb**, be **83.082%** at initialization per process per run.
